### ISSUE TITLE
cartridge pack error allocated memory address 0xffffa81e2010 outside required range

### ISSUE THREAD

**Issue by nmolostvov** [tarantool/tarantool#10502]
**Bug description**

Virtual machine ubuntu 22 - amd64 runned on Mac m3

Place `tarantool --version` output here.

tarantool --version
Tarantool Enterprise 2.11.3-0-ge45691111
Target: Linux-x86_64-Debug
Build options: cmake . -DCMAKE_INSTALL_PREFIX=/__w/sdk/sdk/build.sdk/tarantool-2.11/static-build/tarantool-prefix -DENABLE_BACKTRACE=TRUE
Compiler: GNU-9.3.1
C_FLAGS: -fexceptions -funwind-tables -fasynchronous-unwind-tables -static-libstdc++ -fno-common -fopenmp -msse2 -Wformat -Wformat-security -Werror=format-security -fstack-protector-strong -fPIC -fmacro-prefix-map=/__w/sdk/sdk/tarantool-2.11=. -std=c11 -Wall -Wextra -Wno-gnu-alignof-expression -fno-gnu89-inline -Wno-cast-function-type -Werror
CXX_FLAGS: -fexceptions -funwind-tables -fasynchronous-unwind-tables -static-libstdc++ -fno-common -fopenmp -msse2 -Wformat -Wformat-security -Werror=format-security -fstack-protector-strong -fPIC -fmacro-prefix-map=/__w/sdk/sdk/tarantool-2.11=. -std=c++11 -Wall -Wextra -Wno-invalid-offsetof -Wno-gnu-alignof-expression -Wno-cast-function-type -Werror

**Steps to reproduce**

cartridge pack tgz --spec extractor-scm-1.rockspec --version 0.0.6
   • Packing extractor into tgz
   • Temporary directory is set to /home/**/.cartridge/tmp/pack-0ozby0wppr
   • Initialize application dir
   • git not found. It is possible that some of the extra files normally ignored are shipped to the resulting package. 
   • Build application in /home/**/.cartridge/tmp/pack-0ozby0wppr/package-files/extractor
   • Missed recommended binaries cmake, git, unzip, gcc
   • Running `cartridge.pre-build`
   • Running `tarantoolctl rocks make extractor-scm-1.rockspec`
Missing dependencies for extractor scm-1:
   checks 3.3.0-1 (not installed)
   cartridge 2.12.1-1 (not installed)
   metrics 1.1.0-1 (not installed)
   cartridge-metrics-role 0.1.1-1 (not installed)
   cartridge-cli-extensions 1.1.1-1 (not installed)
   space-explorer 1.1.8-1 (not installed)

extractor scm-1 depends on tarantool (2.11.3-1 provided by VM)
extractor scm-1 depends on lua >= 5.1 (5.1-1 provided by VM)
extractor scm-1 depends on checks 3.3.0-1 (not installed)
Installing /home/**/extractor/sdk/rocks/checks-3.3.0-1.all.rock
No existing manifest. Attempting to rebuild...

checks 3.3.0-1 depends on lua >= 5.1 (5.1-1 provided by VM)
checks 3.3.0-1 is now installed in /home/**/.cartridge/tmp/pack-0ozby0wppr/package-files/extractor/.rocks (license: MIT)

extractor scm-1 depends on cartridge 2.12.1-1 (not installed)
Installing http://rocks.tarantool.org/cartridge-2.12.1-1.all.rock
LuaJIT ASSERT ./tarantool/third_party/luajit/src/lj_gc.c:884: lj_mem_realloc: allocated memory address 0xffffa81e2010 outside required range
   ⨯ Failed to install rocks: Failed to run 
/home/**/extractor/sdk/tarantoolctl rocks make extractor-scm-1.rockspec

**Comment by igormunkin** [comment]
@Buristan, as a wild guess, I believe this is the issue with virtualization layer: the system yields ARM64-specific 48-bit memory address, but LuaJIT expects 47-bit one. It looks like the trick with memory segments, implemented for lightUD several years ago, has to be applied for GC memory allocation either.

Though, I didn't dig the issue at all, so I might be totally wrong here and waiting for you thoughts.

**Comment by Buristan** [comment]
The issue looks similar to LuaJIT/LuaJIT#671. The [patch with a fix for arm64](https://github.com/LuaJIT/LuaJIT/commit/67dbec82) is backported to the tarantool via the e42d116f37578dc76168eefb90eee21cfed2f19a. But the patch uses nomove `mremap()` (i.e., without the `MREMAP_MAYMOVE` flag) only for ARM64 targets. As Igor said, it is possible that the issue is still observed for the non-aarch64 VM running on the aarch64 host because LuaJIT thinks that it is an amd64 host.

I add [experimental patch](https://github.com/tarantool/luajit/commit/48be1cc35ad2190dcb2e5e8f891b042b47d39333) to check this hypothesis, see [the corresponding branch](https://github.com/tarantool/luajit/tree/experimental/mremap-always-nomove). Please apply it to the LuaJIT used in your build and check if the problem is still reproducible.

**Comment by igormunkin** [comment]
> The issue looks similar to [LuaJIT/LuaJIT#671](https://github.com/LuaJIT/LuaJIT/issues/671). 

Yeah, this one looks more relevant to the idea I wrote above; lightUD segments hack is a different story.

> The [patch with a fix for arm64](https://github.com/LuaJIT/LuaJIT/commit/67dbec82) is backported to the tarantool via the [e42d116](https://github.com/tarantool/tarantool/commit/e42d116f37578dc76168eefb90eee21cfed2f19a). But the patch uses nomove `mremap()` (i.e., without the `MREMAP_MAYMOVE` flag) only for ARM64 targets. As Igor said, it is possible that the issue is still observed for the non-aarch64 VM running on the aarch64 host because LuaJIT thinks that it is an amd64 host.

Precisely.
